### PR TITLE
ACCUMULO-4642 Changes to remove ServerConfiguration.getInstance()

### PR DESCRIPTION
**Pull request by milleruntime**
A first step in the bigger refactoring effort for [ACCUMULO-4050](https://issues.apache.org/jira/browse/ACCUMULO-4050).  These changes remove ServerConfiguration.getInstance() and make more use of AccumuloServerContext across TableBalancer classes.  Also deprecates TabletBalancer.init(ServerConfigurationFactory conf) replacing with init(AccumuloServerContext context) to assist in the configuration changes.  

Even without the completion of ACCUMULO-4050, these changes will help simplify the configuration internals of Accumulo.

### CODE REVIEW

File: server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java [hunk 1]
```diff
@@ -71,7 +71,7 @@ public void setUp() {
     opts = new Opts();
     systemConfig = createSystemConfig();
     ServerConfigurationFactory factory = createMock(ServerConfigurationFactory.class);
-    expect(factory.getInstance()).andReturn(instance).anyTimes();
+    // expect(factory.getInstance()).andReturn(instance).anyTimes();
```
Comment (joshelser) [inline server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java]: Remove it.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/TabletBalancer.java [hunk 2]
```diff
@@ -55,16 +56,25 @@
...
   private static final Logger log = LoggerFactory.getLogger(TabletBalancer.class);
 
-  protected ServerConfigurationFactory configuration;
-
   protected AccumuloServerContext context;
 
   /**
    * Initialize the TabletBalancer. This gives the balancer the opportunity to read the configuration.
+   *
+   * @deprecated since 2.0.0; use {@link #init(AccumuloServerContext)} instead.
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/TabletBalancer.java]: Nice javadoc

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java [hunk 3]
```diff
@@ -162,7 +163,16 @@ public long balance(SortedMap<TServerInstance,TabletServerStatus> current, Set<K
     return 100;
   }
 
+  /**
+   * Initialize the TabletBalancer. This gives the balancer the opportunity to read the configuration.
```
Comment (joshelser) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java]: Don't need to duplicate the javadoc here. Most IDEs will inherit it from TabletBalancer

File: server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java [hunk 4]
```diff
@@ -169,23 +170,25 @@ public void span(RemoteSpan s) throws TException {
         if (timeMutation != null)
           writer.addMutation(timeMutation);
       } catch (MutationsRejectedException exception) {
-        log.warn("Unable to write mutation to table; discarding span. set log level to DEBUG for span information and stacktrace. cause: " + exception);
+        log.warn("Unable to write mutation to table; discarding span. set log level to DEBUG for span information and stacktrace. cause: {}", exception,
```
Comment (joshelser) [inline server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java]: While these log message changes are fine, can you please separate them into another JIRA issue?

File: server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java (hunk 1 above)
Comment (milleruntime) [inline server/gc/src/test/java/org/apache/accumulo/gc/SimpleGarbageCollectorTest.java]: Good eye, thanks.

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java (hunk 3 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java]: Agreed.

File: server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java (hunk 4 above)
Comment (milleruntime) [inline server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java]: I pulled these changes from a patch supplied by Christopher and I think some of these got snagged with them.  I guess I can create a ticket with something like "utilize slf4j"?  A quick search in JIRA didn't reveal any tickets like that...

Comment (joshelser) [inline server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java]: Yeah, I'd appreciate just creating one for yourself. It's much easier to cherry-pick the changes later and easier on the eyes now.

Comment (milleruntime) [inline server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java]: No problem, will do.  I now remember grep'ing the code for occurrences of log string concatenation that could be replaced by slf4j logging but quickly gave up, haha.

**Comment by joshelser** [comment]
Cool beans. Thanks for the quick turn-around, Mike. Looks fine to me otherwise.

File: server/base/src/main/java/org/apache/accumulo/server/conf/TableConfWatcher.java [hunk 5]
```diff
@@ -33,11 +33,13 @@
   private static final Logger log = Logger.getLogger(TableConfWatcher.class);
   private final Instance instance;
   private final String tablesPrefix;
+  private final int tablesPrefixLength;
   private ServerConfigurationFactory scf;
 
   TableConfWatcher(Instance instance) {
     this.instance = instance;
     tablesPrefix = ZooUtil.getRoot(instance) + Constants.ZTABLES + "/";
+    tablesPrefixLength = tablesPrefix.length();
```
Comment (ctubbsii) [inline server/base/src/main/java/org/apache/accumulo/server/conf/TableConfWatcher.java]: I think this got carried in from my initial patch that I had attached to the JIRA that @milleruntime picked up. I think it was part of an unrelated change that got picked up in the diff. I actually discarded that change from my workspace because I couldn't remember why I had originally done this. :smile:

File: server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java (hunk 4 above)
Comment (ctubbsii) [inline server/tracer/src/main/java/org/apache/accumulo/tracer/TraceServer.java]: Yeah, sorry about that. The patch I attached to the JIRA included some other changes I had made in my workspace which were unrelated.

File: server/base/src/main/java/org/apache/accumulo/server/conf/TableConfWatcher.java (hunk 5 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/conf/TableConfWatcher.java]: Fixed with 1598590

File: server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java (hunk 3 above)
Comment (milleruntime) [inline server/base/src/main/java/org/apache/accumulo/server/master/balancer/ChaoticLoadBalancer.java]: Fixed with d879158

**Comment by milleruntime** [comment]
These changes ran successfully yesterday with Sunny ITs on my laptop.  Since it looks like Jenkins is still borked, I will just go ahead with pushing these changes today.
